mino/minogrpc/server.go

```diff
@@ -47,384 +38,289 @@ var (
-	// contains elements of kind *x509.Certificate
-	nodesCerts *sync.Map
-
-	routingFactory routing.Factory
+type overlayServer struct {
+	overlay
 
 	handlers map[string]mino.Handler
-
-	traffic *traffic
+	closer   *sync.WaitGroup
 }
 
-// RPC represents an RPC that has been registered by a client, which allows
-// clients to call an RPC that will execute the provided handler.
-//
-// - implements mino.RPC
-type RPC struct {
-	encoder        encoding.ProtoMarshaler
-	handler        mino.Handler
-	srv            *Server
-	uri            string
-	routingFactory routing.Factory
-}
-
-// Call implements mino.RPC. It calls the RPC on each provided address.
-func (rpc *RPC) Call(ctx context.Context, req proto.Message,
-	players mino.Players) (<-chan proto.Message, <-chan error) {
-
-	out := make(chan proto.Message, players.Len())
-	errs := make(chan error, players.Len())
-
-	m, err := ptypes.MarshalAny(req)
-	if err != nil {
-		errs <- xerrors.Errorf("failed to marshal msg to any: %v", err)
-		return out, errs
+// Call implements minogrpc.OverlayClient. It processes the request with the
+// targeted handler if it exists, otherwise it returns an error.
+func (o overlayServer) Call(ctx context.Context, msg *Message) (*Message, error) {
+	// We fetch the uri that identifies the handler in the handlers map with the
+	// grpc metadata api. Using context.Value won't work.
+	headers, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return nil, xerrors.Errorf("header not found in provided context")
 	}
 
-	sendMsg := &Envelope{
-		Message: m,
+	apiURI := headers[headerURIKey]
+	if len(apiURI) == 0 {
+		return nil, xerrors.Errorf("'%s' not found in context header", headerURIKey)
 	}
 
-	go func() {
-		iter := players.AddressIterator()
-		for iter.HasNext() {
-			addrStr := iter.GetNext().String()
-
-			clientConn, err := getConnectionTo(address{addrStr}, rpc.srv.cert,
-				rpc.srv.nodesCerts)
-			if err != nil {
-				errs <- xerrors.Errorf("failed to get client conn for '%s': %v",
-					addrStr, err)
-				continue
-			}
-			cl := NewOverlayClient(clientConn)
+	// If several are provided, only the first one is taken in account.
+	handler, ok := o.handlers[apiURI[0]]
+	if !ok {
+		return nil, xerrors.Errorf("handler '%s' is not registered", apiURI[0])
+	}
 
-			header := metadata.New(map[string]string{headerURIKey: rpc.uri})
-			newCtx := metadata.NewOutgoingContext(ctx, header)
+	message, err := o.encoder.UnmarshalDynamicAny(msg.GetPayload())
+	if err != nil {
+		return nil, xerrors.Errorf("couldn't unmarshal message: %v", err)
+	}
 
-			callResp, err := cl.Call(newCtx, sendMsg)
-			if err != nil {
-				errs <- xerrors.Errorf("failed to call client '%s': %v", addrStr, err)
-				continue
-			}
+	from := o.routingFactory.GetAddressFactory().FromText(msg.GetFrom())
 
-			resp, err := rpc.encoder.UnmarshalDynamicAny(callResp.Message)
-			if err != nil {
-				errs <- xerrors.Errorf("couldn't unmarshal message: %v", err)
-				continue
-			}
+	req := mino.Request{
```

Comment:
I am suddenly wondering: why don't we use the same type of return for streaming? Right now its 
```
Recv(context.Context) (Address, proto.Message, error)
```

but shouldn't it be 

```
Recv(context.Context) (Request, error)
```
 ?

Reply:
As discussed, will do after the PR.